src/pip/_internal/resolution/resolvelib/base.py

```diff
@@ -49,14 +51,19 @@ def __and__(self, other):
             return NotImplemented
         specifier = self.specifier & other.specifier
         hashes = self.hashes & other.hashes(trust_internet=False)
-        return Constraint(specifier, hashes)
+        links = self.links
+        if other.link and other.link not in links:
+            links += (other.link,)
```

Comment:
… and change this to something like `links = [other.link, *links]` (hooray Python 3).

Reply:
Thinking about this more, maybe we should use `Set[Link]` instead? This will save some work for the resolver when there are duplicate links in the constraints files.